prac_07/project_management.py

```diff
@@ -0,0 +1,152 @@
+             "- (F)ilter projects by date\n"
+             "- (A)dd new project\n"
+             "- (U)pdate project\n"
+             "- (Q)uit")
+DEFAULT_FILENAME = "projects.txt"
+NAME_INDEX = 0
+START_DATE_INDEX = 1
+PRIORITY_INDEX = 2
+COST_ESTIMATE_INDEX = 3
+COMPLETION_PERCENTAGE_INDEX = 4
+
+
+def main():
+    """Allow for a file to be loaded or saved to, display and filter the projects, and also add or update projects."""
+    projects = process_file(DEFAULT_FILENAME, NAME_INDEX, START_DATE_INDEX, PRIORITY_INDEX, COST_ESTIMATE_INDEX,
+                            COMPLETION_PERCENTAGE_INDEX)
+    print(MENU_TEXT)
+    menu_selection = input(">>> ").upper()
+    while menu_selection != "Q":
+        if menu_selection == "L":
+            new_filename = input("File name: ")
+            try:
+                projects = process_file(new_filename, NAME_INDEX, START_DATE_INDEX, PRIORITY_INDEX, COST_ESTIMATE_INDEX,
+                                        COMPLETION_PERCENTAGE_INDEX)
+            except FileNotFoundError:
+                print(f"Could not find file {new_filename}")
+        elif menu_selection == "S":
+            new_filename = input("File name: ")
+            write_to_file(projects, new_filename)
+        elif menu_selection == "D":
+            display_projects(projects)
+        elif menu_selection == "F":
+            filter_by_date(projects)
+        elif menu_selection == "A":
+            print("Let's add a new project")
+            new_project = get_new_project()
+            projects.append(new_project)
+        elif menu_selection == "U":
+            projects = update_project(projects)
+            print(projects)
+        else:
+            print("Invalid selection.")
+        print(MENU_TEXT)
+        menu_selection = input(">>> ").upper()
+    write_to_file(projects, DEFAULT_FILENAME)
+    print("Thank you for using custom-built project management software.")
+
+
+def display_projects(projects):
+    """Display projects grouped based on completion percentage."""
+    projects.sort()
+    print("Incomplete projects:")
+    for project in projects:
+        if project.completion_percentage < 100.0:
+            print(project)
+    print("Completed projects:")
+    for project in projects:
+        if project.completion_percentage == 100.0:
+            print(project)
+
+
+def process_file(filename, name_index, start_date_index, priority_index, cost_estimate_index,
+                 completion_percentage_index):
+    """Process data from file to store in a list of objects."""
+    projects = []
+    with open(filename, "r") as in_file:
+        in_file.readline()  # ignore headings
+        for line in in_file:
+            parts = line.strip().split("\t")
+            project = Project(parts[name_index], parts[start_date_index], int(parts[priority_index]),
+                              float(parts[cost_estimate_index]), float(parts[completion_percentage_index]))
+            projects.append(project)
```

Comment:
Does using constants for indices make the code more readable ?
Do you need to create a new variable to append the project ?